### PR TITLE
impr: use tsrest/zod for type-safety on both ends (@fehmer)

### PR DESCRIPTION
**Pull request by fehmer**
Ensure type-safety between frontend and backend by using [ts-rest](https://ts-rest.com/) and [zod](https://zod.dev/) (#5440)

Features: 
- define path, http method, request body, query/path params and the response codes/types in the contract
- define typescript types from the contract
- replaces the need for the SharedTypes 
- generated openApi documentation `/v2/docs` and `/v2/docs.json`
- backend 
  - validate request (and response optionally) from the contract schema
  - get compile error if you forgot to implement an endpoint
  - get typed request query, body and path params on the controller
  - return typed response on the controller
  - fewer files to touch when adding a new top level route
  - using the same error handling (`throw MonkeyError`) as before
- frontend
  -  typed request, query, path and responses including status codes
  -  frontend validation using the same schema
  - removed the wrapping of the api so we can directly use the types (see changes in `db.ts`)
  - keep the same error handling as before.

### CODE REVIEW

File: backend/src/api/routes/swagger.ts [hunk 1]
```diff
@@ -31,10 +29,38 @@ function addSwaggerMiddlewares(app: Application): void {
...
+        termsOfService: "https://monkeytype.com/terms-of-service",
+        contact: {
+          name: "Support",
+          email: "support@monkeytype.com",
+        },
+      },
+    },
+    { jsonQuery: true, setOperationId: "concatenated-path" }
+  );
+  app.use("/v2/docs.json", (req, res) => {
```
Comment (fehmer) [inline backend/src/api/routes/swagger.ts:56]: for now we'll serve two documentations, the old manual maintained one and the generated one for ts-rest endpoints. I tried merging both contracts but they are using different openapi versions.

File: backend/src/api/schemas/config-schema.ts [hunk 2]
```diff
@@ -11,6 +11,7 @@ const CARET_STYLES = [
   "banana",
 ];
 
+//TODO replaced
```
Comment (fehmer) [inline backend/src/api/schemas/config-schema.ts:14]: currently still used in profiles, but we can remove this when we move profiles to the new approach

File: backend/src/dal/config.ts [hunk 3]
```diff
@@ -22,34 +23,34 @@ const configLegacyProperties = [
...
-      { upsert: true }
-    );
+  return await getConfigCollection().updateOne(
+    { uid },
+    { $set: configChanges, $unset: unset },
+    { upsert: true }
+  );
 }
 
 export async function getConfig(
```
Comment (fehmer) [inline backend/src/dal/config.ts:47]: fixed the type, we are using `{_id, uid, config: { ....}`

File: backend/src/middlewares/auth.ts [hunk 4]
```diff
@@ -27,80 +29,106 @@ const DEFAULT_OPTIONS: RequestAuthenticationOptions = {
   requireFreshToken: false,
 };
 
-function authenticateRequest(authOptions = DEFAULT_OPTIONS): Handler {
```
Comment (fehmer) [inline backend/src/middlewares/auth.ts:30]: all the changes are git having problems. I renamed the old method to `_authenticateRequestInternal`

File: backend/src/utils/monkey-response.ts [hunk 5]
```diff
@@ -36,3 +41,29 @@ export function handleMonkeyResponse(
...
+  }
+
+  public static fromDB<T extends { _id: string }>(
+    message: string,
+    data: MonkeyTypes.WithObjectId<T>,
+    status = 200
+  ): MonkeyResponse2<T> {
+    return new MonkeyResponse2(
+      message,
+      { _id: data._id.toString(), ...omit(data, "_id") } as unknown as T,
```
Comment (fehmer) [inline backend/src/utils/monkey-response.ts:65]: In the DB we use `ObjectId` and  in rest we use `string` so we need to convert.

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/9941977475

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/9945772665

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/9945995908

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/9946331932

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/9955434341

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/9955477960

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/9957502667

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/9957502667

**Pull request merged**